File: src/main/java/io/github/hdzitao/editstarters/dependency/Points.java
package io.github.hdzitao.editstarters.dependency;

import java.util.Collection;
import java.util.Objects;

/**
 * Point方法集
 *
 * @version 3.2.0
 */
public final class Points {
    private Points() {
    }

    /**
     * 查找
     */
    public static <P extends Point> P find(Collection<P> points, Point point) {
        for (P p : points) {
            if (Objects.equals(point.point(), p.point())) {
                return p;
            }
        }

        return null;
    }

    /**
     * 包含
     */
    public static boolean contains(Collection<? extends Point> points, Point point) {
        return find(points, point) != null;
    }

    /**
     * 不重复添加
     */
    public static <P extends Point> void addUniq(Collection<P> points, P point) {
        if (point != null && !contains(points, point)) {
            points.add(point);
        }
    }

    /**
     * 不重复添加
     */
    public static <P extends Point> void addAllUniq(Collection<P> points, Collection<P> others) {
        for (P o : others) {
            addUniq(points, o);
        }
    }
}

File: src/main/kotlin/io/github/hdzitao/editstarters/dependency/points.kt
package io.github.hdzitao.editstarters.dependency

/**
 * 查找
 */
fun <T : Point> Collection<T>.findPont(point: Point) = this.find { it.point() == point.point() }

/**
 * 包含
 */
fun Collection<Point>.hasPoint(point: Point) = findPont(point) != null

/**
 * 不重复添加
 */
fun <T : Point> MutableCollection<T>.addPointUniq(point: T?) {
    if (point != null && !hasPoint(point)) {
        add(point)
    }
}

/**
 * 不重复添加
 */
fun <T : Point> MutableCollection<T>.addAllPointsUniq(others: Collection<T>) {
    for (o in others) {
        addPointUniq(o)
    }
}